File: 94 inorderTraversal/94(1).cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int>res;
        if(root==nullptr){
            return res;
        }
        TreeNode* node=root;
        stack<TreeNode*>stk;
        while(!stk.empty()||node!=nullptr){
            while(node!=nullptr){
                stk.push(node);
                node=node->left;
            }
            node=stk.top();
            stk.pop();
            res.push_back(node->val);
            node=node->right;
        }
        return res;
    }
};

File: 94 inorderTraversal/94(1).py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        stk=[]
        res=[]
        node = root
        while node or stk:
            while node:
                stk.append(node)
                node=node.left
            node=stk.pop()
            res.append(node.val)
            node=node.right
        return res